sdk_version_resolver.go

```diff
@@ -0,0 +1,156 @@
+package dotnetcoresdk
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/BurntSushi/toml"
+	"github.com/Masterminds/semver"
+	"github.com/paketo-buildpacks/packit"
+	"github.com/paketo-buildpacks/packit/postal"
+)
+
+type RuntimeToSdks struct {
+	RuntimeVersion string   `toml:"runtime-version"`
+	SDKs           []string `toml:"sdks"`
+}
+
+type SDKVersionResolver struct {
+	logger LogEmitter
+}
+
+func NewSDKVersionResolver(logger LogEmitter) SDKVersionResolver {
+	return SDKVersionResolver{logger: logger}
+}
+
+func (r SDKVersionResolver) Resolve(path string,
+	entry packit.BuildpackPlanEntry,
+	stack string) (postal.Dependency, error) {
+
+	var sdkVersion string
+	if sdkVersionStruct, ok := entry.Metadata["version"]; ok {
+		sdkVersion = sdkVersionStruct.(string)
+	}
+
+	if sdkVersion == "" || sdkVersion == "default" {
+		sdkVersion = "*"
+	}
+
+	sdkConstraint, err := semver.NewConstraint(sdkVersion)
+	if err != nil {
+		return postal.Dependency{}, err
+	}
+
+	availableDependencies, err := gatherDependenciesFromBuildpackTOML(path, "dotnet-sdk", stack) // get the dependency from the buildpack.toml and return it
+	if err != nil {
+		return postal.Dependency{}, fmt.Errorf("buildpack.toml could not be parsed: %w", err)
+	}
+
+	var sdkDependencies []postal.Dependency
+	for _, dep := range availableDependencies {
+		depVersion, err := semver.NewVersion(dep.Version)
+		if err != nil {
+			return postal.Dependency{}, err
+		}
+		if sdkConstraint.Check(depVersion) {
+			sdkDependencies = append(sdkDependencies, dep)
+		}
+	}
+
+	if len(sdkDependencies) == 0 {
+		var supportedVersions []string
+		for _, dependency := range availableDependencies {
+			supportedVersions = append(supportedVersions, dependency.Version)
+		}
+		// panic("requested version not available")
+		return postal.Dependency{}, fmt.Errorf(
+			"failed to satisfy %q dependency for stack %q with version constraint %q: no compatible versions. Supported versions are: [%s]",
+			entry.Name,
+			stack,
+			sdkVersion,
+			strings.Join(supportedVersions, ", "),
+		)
+	}
+	if runtimeVersion := os.Getenv("RUNTIME_VERSION"); runtimeVersion != "" {
```

Comment:
Since we check this env var set by the Dotnet Runtime buildpack, we should explicitly require `dotnet-runtime` at build time.